lazy_dataset/core.py

```diff
@@ -2313,6 +2450,102 @@ def __iter__(self):
             print(f'Dropped {dropped_count} examples')
 
 
+class CacheDataset(Dataset):
+    def __init__(self, dataset: Dataset, keep_mem_free=None,
+                 immutable_warranty: str = 'pickle') -> None:
+        super().__init__()
+        assert dataset.indexable, (
+            'CacheDataset only works if dataset is indexable!'
+        )
+        self.dataset = dataset
+        self.cache = {}
+
+        self._keep_mem_free = self._get_memory_size(keep_mem_free)
+        self.immutable_warranty = immutable_warranty
+
+        if immutable_warranty == 'pickle':
+            self._serialize = pickle.dumps
+            self._deserialize = pickle.loads
+        elif immutable_warranty == 'copy':
+            self._serialize = lambda x: x
+            self._deserialize = deepcopy
+        else:
+            raise ValueError(immutable_warranty)
+
+    @property
+    def indexable(self) -> bool:
+        return self.dataset.indexable
+
+    @property
+    def ordered(self) -> bool:
+        return self.dataset.ordered
+
+    def keys(self) -> list:
+        return self.dataset.keys()
+
+    @staticmethod
+    def _get_memory_size(keep_mem_free):
+        if keep_mem_free is None:
+            return None
+
+        if keep_mem_free.strip().endswith('%'):
+            import psutil
+            value = float(keep_mem_free.strip(' %'))
+            assert 0 <= value <= 100, value
+            return psutil.virtual_memory().total * value / 100
+        else:
+            import humanfriendly
+            return humanfriendly.parse_size(keep_mem_free, binary=True)
+
+    def __getitem__(self, item):
+        if isinstance(item, numbers.Integral):
+            item = self.keys()[item]
+
+        if isinstance(item, str):
+            if item not in self.cache:
+                # Check if we have enough free memory
+                if self._keep_mem_free is not None:
+                    import psutil
+                    if psutil.virtual_memory().available <= self._keep_mem_free:
+                        # Return without writing to cache if there is not enough
+                        # free memory
+                        return self.dataset[item]
+
+                self.cache[item] = self._serialize(self.dataset[item])
+            return self._deserialize(self.cache[item])
+        else:
+            # Support for slices etc.
+            return super().__getitem__(item)
+
+    def __iter__(self):
+        for k in self.keys():
+            print(k)
```

Comment:
Can you remove that print?

Reply:
Sorry, the new git commit dialog in PyCharm is confusing... It's removed now